File: archiso/airootfs/usr/share/churros/services/battery.py
import subprocess


class BatteryService:

    @staticmethod
    def _run(cmd):

        try:

            return subprocess.run(

                cmd,

                capture_output=True,
                text=True,
                timeout=3

            ).stdout

        except Exception:

            return None

    @staticmethod
    def _find_battery():

        out = BatteryService._run(["upower", "-e"])

        if out is None:

            return None

        for line in out.splitlines():

            if "battery" in line.lower():

                return line

        return None

    @staticmethod
    def available():

        return BatteryService._find_battery() is not None

    @staticmethod
    def get():

        battery = BatteryService._find_battery()

        if battery is None:

            return {"available": False}

        info_out = BatteryService._run(["upower", "-i", battery])

        if info_out is None:

            return {"available": False}

        info = {

            "available": True,
            "percentage": 0,

            "state": "unknown",

            "time_to_full": "",
            "time_to_empty": "",

            "icon": "󰂎",
            "charging_icon": ""
        }

        for line in info_out.splitlines():

            line = line.strip()

            if line.startswith("state:"):

                info["state"] = line.split(":", 1)[1].strip()

            elif line.startswith("percentage:"):

                try:

                    info["percentage"] = int(
                        line.split(":", 1)[1]
                        .replace("%", "")
                        .strip()
                    )

                except ValueError:
                    pass

            elif line.startswith("time to full:"):

                info["time_to_full"] = line.split(":", 1)[1].strip()

            elif line.startswith("time to empty:"):

                info["time_to_empty"] = line.split(":", 1)[1].strip()

        p = info["percentage"]

        charging = info["state"] in ("charging", "fully-charged", "pending-charge")

        if charging:
            if p < 10:
                info["icon"] = "󰢜"
            elif p < 30:
                info["icon"] = "󰂆"
            elif p < 50:
                info["icon"] = "󰂇"
            elif p < 70:
                info["icon"] = "󰂈"
            elif p < 85:
                info["icon"] = "󰢝"
            elif p < 95:
                info["icon"] = "󰂉"
            else:
                info["icon"] = "󰂊"
        else:
            if p >= 95:
                info["icon"] = "󰁹"
            elif p >= 80:
                info["icon"] = "󰂂"
            elif p >= 60:
                info["icon"] = "󰂀"
            elif p >= 40:
                info["icon"] = "󰁾"
            elif p >= 20:
                info["icon"] = "󰁼"
            else:
                info["icon"] = "󰂎"

        return info

File: rust/services/src/battery.rs
// Servicio de batería vía upower (equivalente a services/battery.py)

use crate::run;

fn find_battery() -> Option<String> {
    let out = run(&["upower", "-e"], 3000)?;
    out.1
        .lines()
        .find(|l| l.to_lowercase().contains("battery"))
        .map(|l| l.to_string())
}

fn battery_icon(percentage: u8, charging: bool) -> &'static str {
    if charging {
        if percentage < 10 {
            "󰢜"
        } else if percentage < 30 {
            "󰂆"
        } else if percentage < 50 {
            "󰂇"
        } else if percentage < 70 {
            "󰂈"
        } else if percentage < 85 {
            "󰢝"
        } else if percentage < 95 {
            "󰂉"
        } else {
            "󰂊"
        }
    } else if percentage >= 95 {
        "󰁹"
    } else if percentage >= 80 {
        "󰂂"
    } else if percentage >= 60 {
        "󰂀"
    } else if percentage >= 40 {
        "󰁾"
    } else if percentage >= 20 {
        "󰁼"
    } else {
        "󰂎"
    }
}

#[derive(Debug, Clone)]
pub struct BatteryInfo {
    pub available: bool,
    pub percentage: u8,
    pub state: String,
    pub time_to_full: String,
    pub time_to_empty: String,
    pub icon: String,
}

pub fn available() -> bool {
    find_battery().is_some()
}

pub fn get() -> BatteryInfo {
    let mut info = BatteryInfo {
        available: false,
        percentage: 0,
        state: "unknown".to_string(),
        time_to_full: String::new(),
        time_to_empty: String::new(),
        icon: "󰂎".to_string(),
    };

    let Some(battery) = find_battery() else {
        return info;
    };
    let Some((_, out, _)) = run(&["upower", "-i", &battery], 3000) else {
        return info;
    };

    info.available = true;

    for line in out.lines() {
        let line = line.trim();
        if let Some(v) = line.strip_prefix("state:") {
            info.state = v.trim().to_string();
        } else if let Some(v) = line.strip_prefix("percentage:") {
            info.percentage = v
                .replace('%', "")
                .trim()
                .parse()
                .unwrap_or(info.percentage);
        } else if let Some(v) = line.strip_prefix("time to full:") {
            info.time_to_full = v.trim().to_string();
        } else if let Some(v) = line.strip_prefix("time to empty:") {
            info.time_to_empty = v.trim().to_string();
        }
    }

    let charging = matches!(
        info.state.as_str(),
        "charging" | "fully-charged" | "pending-charge"
    );
    info.icon = battery_icon(info.percentage, charging).to_string();

    info
}
